tree/treeplayer/inc/TBranchProxy.h

```diff
@@ -160,10 +168,185 @@ namespace Detail {
+         }
+      }
+
+      Bool_t ReadParentCollectionNoPointer() {
+         auto treeEntry = fDirector->GetReadEntry();
+         if (treeEntry != fRead) {
+            const Bool_t result = fParent->Read();
+            fRead = treeEntry;
+            fCollection->PopProxy(); // works even if no proxy env object was set.
+            fCollection->PushProxy( fWhere );
+            return result;
+         } else {
+            return IsInitialized();
+         }
+      }
+
+      Bool_t ReadParentCollectionPointer() {
+         auto treeEntry = fDirector->GetReadEntry();
+         if (treeEntry != fRead) {
+            const Bool_t result = fParent->Read();
+            fRead = treeEntry;
+            fCollection->PopProxy(); // works even if no proxy env object was set.
+            fCollection->PushProxy( *(void**)fWhere );
+            return result;
+         } else {
+            return IsInitialized();
+         }
+      }
+
+      Bool_t ReadNoParentNoBranchCountCollectionPointer() {
+         auto treeEntry = fDirector->GetReadEntry();
+         if (treeEntry != fRead) {
+            Bool_t result = (-1 != fBranch->GetEntry(treeEntry));
+            fRead = treeEntry;
+            fCollection->PopProxy(); // works even if no proxy env object was set.
+            fCollection->PushProxy( *(void**)fWhere );
+            return result;
+         } else {
+            return IsInitialized();
+         }
+      }
+
+      Bool_t ReadNoParentNoBranchCountCollectionNoPointer() {
+         auto treeEntry = fDirector->GetReadEntry();
+         if (treeEntry != fRead) {
+            Bool_t result = (-1 != fBranch->GetEntry(treeEntry));
+            fRead = treeEntry;
+            fCollection->PopProxy(); // works even if no proxy env object was set.
+            fCollection->PushProxy( fWhere );
+            return result;
+         } else {
+            return IsInitialized();
+         }
+      }
+
+      Bool_t ReadNoParentNoBranchCountNoCollection() {
+         auto treeEntry = fDirector->GetReadEntry();
+         if (treeEntry != fRead) {
+            Bool_t result = (-1 != fBranch->GetEntry(treeEntry));
+            fRead = treeEntry;
+            return result;
+         } else {
+            return IsInitialized();
+         }
+      }
+
+      Bool_t ReadNoParentBranchCountCollectionPointer() {
+         auto treeEntry = fDirector->GetReadEntry();
+         if (treeEntry != fRead) {
+            Bool_t result = (-1 != fBranchCount->GetEntry(treeEntry));
+            result &= (-1 != fBranch->GetEntry(treeEntry));
+            fRead = treeEntry;
+            fCollection->PopProxy(); // works even if no proxy env object was set.
+            fCollection->PushProxy( *(void**)fWhere );
+            return result;
+         } else {
+            return IsInitialized();
+         }
+      }
+
+      Bool_t ReadNoParentBranchCountCollectionNoPointer() {
+         auto treeEntry = fDirector->GetReadEntry();
+         if (treeEntry != fRead) {
```

Comment:
There is a lot of repetition of this `if` condition. This violates "don't repeat yourself"... Maybe it could be factored out into a separate function.

Reply:
This is performance 'critical' code.  It is not clear whether saving the repeated boiler plate (the outer if) is worth the cost (having a function template taking a function as an argument) in term of readbility ....

Reply:
It is not clear to me the code with repetitions is faster than using some inline functions. I don't think you need a template taking a function as argument at all. Also, `ReadNoParentBranchCountCollectionNoPointer()` is already quite a handful in terms of readability. I'm not sure what it's supposed to mean/do. Anyway, below is an example of how to reduce code duplication:

```cpp
  inline bool UpdateTreeEntry() {
    if ((auto entry = fDirector->GetReadEntry()) != fRead) {
      fRead = entry;
      return true;
    } else {
      return false;
    }
  }

  inline bool ReadBranch() {
    return fBranch->GetEntry(fRead) != -1;
  }

  inline bool ReadBranchCount() {
    return fBranch->GetEntry(fRead) != -1;
  }

  inline void UpdateCollectionProxy() {
    fCollection->PopProxy(); // works even if no proxy env object was set.
    fCollection->PushProxy( fWhere );  
  }

  bool ReadNoParentBranchCountNoCollection() {
    if (!UpdateTreeEntry())
      return IsInitialized();
    return ReadBranch() && ReadBranchCount();
  }

  bool ReadNoParentBranchCountCollectionNoPointer() {
    bool ret = ReadNoParentBranchCountNoCollection();
    UpdateCollectionProxy();
    return ret;
  }
```

Reply:
> Also, ReadNoParentBranchCountCollectionNoPointer() is already quite a handful in terms of readability. I'm not sure what it's supposed to mean/do. 

The name is lame but a straightforward expression of the unfolding of the combinatoric of the original Read function.

> Anyway, below is an example of how to reduce code duplication:

It does indeed reduce duplication.  It does also make it harder for me to reason about performance.  (it does look like that version as one more branch in some of the cases that the PR and I can't know whether that branch will optimized out or not ...)

Reply:
> Another comment is that making the EReadType a bit mask would allow you to collapse all these strangely named functions into a single one:

This is an interesting option.  Off hand I can not tell whether reintroducing the branches with much faster to evaluate condition will be faster or not than the function pointer ....